### ISSUE TITLE
ICE on assert!("test")

### ISSUE THREAD

**Issue by wesleywiser** [rust-lang/rust#28308]
While playing around on the playground, I found that this reports an ICE:

```rust
fn main() {
    assert!("test");
}
```

```
<std macros>:2:4: 2:19 error: cannot apply unary operator `!` to type `&'static str`
(internal compiler error: unprintable span)
<std macros>:1:1: 5:46 note: in expansion of assert!
<anon>:2:5: 2:21 note: expansion site
error: aborting due to previous error
playpen: application terminated with error code 101
```

This repros on stable, beta, and nightly

**Comment by jld** [comment]
I ran into this by accidentally typing `assert!` instead of `assert_eq!` too early in the morning, so I did some `git grep` and experimenting, and I think I know what's happening here: this is a `SpanLinesError::DistinctSources`, because `assert!` expands to `if !$cond { ... }`, so the compiler wants to complain about `!$cond`, and it sees that that expression starts at the `!` in `libcore/macros.rs` and ends at the end of… the expression given as the `$cond` argument to the macro in the code being compiled.

There's an easier way to see this:
```Rust
#![allow(unused_parens)]
macro_rules! my_assert1 { ($e:expr) => { if !$e { panic!() } } }
macro_rules! my_assert2 { ($e:expr) => { if !($e) { panic!() } } }
// Meaningless text goes here
// And some more meaningless text
fn main() {
   my_assert2!("test");
}
```
This fails nicely:
```
issue28308.rs:3:45: 3:50 error: cannot apply unary operator `!` to type `&'static str`
issue28308.rs:3 macro_rules! my_assert2 { ($e:expr) => { if !($e) { panic!() } } }
                                                            ^~~~~
issue28308.rs:7:4: 7:24 note: in this expansion of my_assert2! (defined in issue28308.rs)
```
But change the last part to `my_assert1!`:
```
issue28308.rs:2:45: 7:22 error: cannot apply unary operator `!` to type `&'static str`
issue28308.rs:2 macro_rules! my_assert1 { ($e:expr) => { if !$e { panic!() } } }
issue28308.rs:3 macro_rules! my_assert2 { ($e:expr) => { if !($e) { panic!() } } }
issue28308.rs:4 // Meaningless text goes here
issue28308.rs:5 // And some more meaningless text
issue28308.rs:6 fn main() {
issue28308.rs:7    my_assert1!("test");
issue28308.rs:7:4: 7:24 note: in this expansion of my_assert1! (defined in issue28308.rs)
```
Alternately, with a macro variable at the start of an offending expression, like this:
```Rust
macro_rules! as_isize { ($e:expr) => { $e as isize } }
fn main() { as_isize!("test"); }
```
…we get the other kind of unprintable span, where the beginning is after the end:
```
issue28308alt.rs:2:23: 1:51 error: casting `&'static str` as `isize` is invalid
(internal compiler error: unprintable span)
```

**Comment by jld** [comment]
It looks like #25385 already discovered that last one.
